File: challenges/project-euler/python3/3.py
# Trying to do it properly this time.

# Check if a number is prime or not.
def is_prime(num):
    for i in range(2, (num // 2) + 1):
        if (num % i == 0): return False
    return True

# Get the next prime number, after a given number.
def next_prime(num):
    while True:
        num += 1
        if is_prime(num): return num

# Find the largest prime factor of a number.
def largest_prime_factor(num):
    next = 2
    while True:
        if is_prime(num): return num
        if num % next == 0: num //= next
        next = next_prime(next)

# What is the largest prime factor of the number 600851475143 ?
print(largest_prime_factor(600851475143))

File: challenges/project-euler/c++/3.cpp
#include <iostream>

// Finds the next prime number.
uint64_t next_prime(uint64_t num)
{
	do
	{
		bool prime = true;
		num++;
		for (unsigned long long i = 2; i < num; i++)
			if (num % i == 0) { prime = false; break; }
		if (prime) return num;
	} while(true);
}

int main()
{
	bool found = false;
	uint64_t num = 600851475143ULL;
	do
	{
		found = true;
		for (uint64_t i = 3; i < num; i = next_prime(i))
			if (num % i == 0) { num /= i; found = false; break; }
	} while(!found);
	std::cout << num << std::endl;
	return 0;
}
